File: 코드/ch5/clip4/boj1759/main.py
import sys

def is_vowel(c):
    return c == 'a' or c == 'e' or c == 'i' or c == 'o' or c == 'u'

def generate(length, index, vowel_cnt):
    if length == l:
        if vowel_cnt >= 1 and l - vowel_cnt >= 2:
            print(''.join(password))
        return
    if index < c:
        password[length] = input[index]
        v = 1 if is_vowel(input[index]) else 0
        generate(length + 1, index + 1, vowel_cnt + v)
        password[length] = 0
        generate(length, index + 1, vowel_cnt)

l, c = map(int, sys.stdin.readline().split())
input = sys.stdin.readline().split()
password = [0] * l
input.sort()
generate(0, 0, 0)

File: 코드/ch5/clip4/boj1759/Main.java
import java.util.Arrays;
import java.util.Scanner;

class Main {
    static int l, c;
    static char[] input;
    static char[] password;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        l = sc.nextInt();
        c = sc.nextInt();

        input = new char[c];
        password = new char[l];
        for (int i = 0; i < c; i++) {
            input[i] = sc.next().charAt(0);
        }
        Arrays.sort(input);
        generate(0, 0, 0);
    }

    static void generate(int length, int index, int vowelCnt) {
        // base case
        if(length == l) {
            if (vowelCnt >= 1 && l - vowelCnt >= 2) {
                System.out.println(password);
            }
            return;
        }

        // recursive case
        if(index < c) {
            // password에 input[index] 사용하는 경우
            password[length] = input[index];
            int v = isVowel(input[index]) ? 1 : 0;
            generate(length + 1, index + 1, vowelCnt + v);

            // password에 input[index] 사용하지 않는 경우
            password[length] = 0;
            generate(length, index + 1, vowelCnt);
        }
    }

    static boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
}